src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.Decode.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Pkcs;
+using System.Security.Cryptography.X509Certificates;
+
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal.OpenSsl
+{
+    internal sealed partial class DecryptorPalOpenSsl : DecryptorPal
+    {
+        public static DecryptorPalOpenSsl Decode(
+            byte[] encodedMessage,
+            out int version,
+            out ContentInfo contentInfo,
+            out AlgorithmIdentifier contentEncryptionAlgorithm,
+            out X509Certificate2Collection originatorCerts,
+            out CryptographicAttributeObjectCollection unprotectedAttributes)
+        {
+            SafeCmsHandle cmsHandle = Interop.Crypto.CmsDecode(encodedMessage, encodedMessage.Length);
+            Interop.Crypto.CheckValidOpenSslHandle(cmsHandle);
+
+            string oid = cmsHandle.GetContentType();
+
+            if (oid != Oids.Pkcs7Enveloped)
+                throw new CryptographicException(SR.Cryptography_Cms_Invalid_Message_Type);
+
+            originatorCerts = cmsHandle.GetOriginatorCerts();
+            contentInfo = cmsHandle.GetEmbeddedContent();
+            RecipientInfoCollection recipientInfos = cmsHandle.GetRecipients();
+
+
+            // DerSequenceReader needed to expose the parts of the message that the OpenSsl library doesn't
+            // expose in their CMS API. 
+            DerSequenceReader encodedCms = new DerSequenceReader(encodedMessage);
+            try
+            {
+                // While the other elements are obtained from interacting from OpenSsl, these values had to
+                // be obtained from parsing the encoding as there seems to be no way of obtaining them through 
+                // their API. As such, don't move these around as they need to be read in this order
+
+                // Read the version, but first skip the OID associated with the message, and open the inner constructed
+                // sequence. 
+                Debug.Assert(encodedCms.PeekTag() == (byte)DerSequenceReader.DerTag.ObjectIdentifier,
+                "Expected to skip an OID while reading version");
```

Comment:
Style: Chopped arguments should be indented to help show they're part of the previous line.

I'm, personally, usually a fan of "chop before first argument".  And almost always have a blank line (or closing curly brace) on the following line.  So I'd have this as

```C#
Debug.Assert(
    encodedCms.PeekTag() == (byte)DerSequenceReader.DerTag.ObjectIdentifier,
    "Expected to skip an OID while reading version");

 encodedCms.SkipValue();
```